sstables/storage.cc

```diff
@@ -36,10 +36,57 @@
 
 namespace sstables {
 
+class dir_wrapper final {
+    std::filesystem::path _pathname;
+    file _file;
+    bool _file_stale;
+
+public:
+    explicit dir_wrapper(std::filesystem::path pathname) : _pathname(std::move(pathname)) {};
+    dir_wrapper(const dir_wrapper&) = delete;
+    dir_wrapper& operator=(const dir_wrapper&) = delete;
+    dir_wrapper(dir_wrapper&&) = delete;
+    dir_wrapper& operator=(dir_wrapper&&) = delete;
+    ~dir_wrapper() = default;
+
+    const std::filesystem::path::string_type& native() const noexcept {
+        return _pathname.native();
+    }
+
+    operator std::filesystem::path() const noexcept {
+        return _pathname;
+    }
+
+    dir_wrapper& operator=(const sstring &nd) {
+        _pathname = nd;
+        _file_stale = true;
+        return *this;
+    }
+
+    future<> sync(const sstable& sst) {
+        if (_file && _file_stale) {
+            co_await sst.sstable_write_io_check(std::mem_fn(&file::close), _file);
+            _file = file();
+        }
+        if (!_file) {
+            _file = co_await sst.sstable_write_io_check(open_directory, _pathname.native());
+            _file_stale = false;
+        }
+        co_await sst.sstable_write_io_check(std::mem_fn(&file::flush), _file);
+    };
+
+    future<> destroy(const sstable& sst) {
```

Comment:
Hmm, not sure about the interface... it seems weird to close the directory in the context of a particular sstable in it, just to get to `sst.sstable_write_io_check`.  If needed, we can implement a new error handler.

Reply:
Also, typically we call such a method `close()`.

Reply:
> Hmm, not sure about the interface... it seems weird to close the directory in the context of a particular sstable in it, just to get to `sst.sstable_write_io_check`. If needed, we can implement a new error handler.

The parameter list was modeled on `storage::destroy`, from commit 81a1416ebf23 ("sstables: Add and call storage::destroy()", 2023-05-02).

> Also, typically we call such a method `close()`.

I took the name as well from commit 81a1416ebf23.

Reply:
I'm not sure what the original thinking behind storage::destroy... @xemul can you please share your insight?

Here, the directory wrapper is a standalone entity, unrelated to a particular sstable.
In theory, we could tie it the an sstable_directory and state (which stands for a subdirectory in the table's base directory), but I'm not sure we need it to be that tightly coupled with the table files hierarchy.

The point is that the dir_wrapper object should be useful for opening the directory on demand, given its path, and flush it multiple times over its lifetime. And finally, allow closing it using an async close() function, before the wrapper is destroyed.

As for assigning a new path to it.
If it's needed only for testing, we don't really need a method for it, just implement the move ctor and assignment operator, and assign a new dir_wrapper onto the existing one.

Reply:
> I'm not sure what the original thinking behind storage::destroy... @xemul can you please share your insight?
> 
> Here, the directory wrapper is a standalone entity, unrelated to a particular sstable. In theory, we could tie it the an sstable_directory and state (which stands for a subdirectory in the table's base directory), but I'm not sure we need it to be that tightly coupled with the table files hierarchy.
> 
> The point is that the dir_wrapper object should be useful for opening the directory on demand, given its path, and flush it multiple times over its lifetime.

The `destroy` change goes back to @xemul's [comment](https://github.com/scylladb/scylladb/pull/19624#discussion_r1670390612). I may have misunderstood that comment of course. Either way, I'm dropping the destroy modification from v5. (Similarly pointed out by @avikivity [here](https://github.com/scylladb/scylladb/pull/19624#discussion_r1681255051).)

Reply:
@bhalevy 

> As for assigning a new path to it.
> If it's needed only for testing, we don't really need a method for it, just implement the move ctor and assignment operator, and assign a new dir_wrapper onto the existing one.

It's not only needed for testing; it's also put to use in `filesystem_storage::move()`.

Anyway, I think to have implemented this in v5.

(I believe that I'll only be able to appreciate the full depth of this change -- i.e., of replacing `_dir` altogether with a new object, upon changing the directory name -- once I get around working on your `_close_fut` proposal, additionally.

Because, in the current state of my code, when the old `_dir` object vanishes, its `_file` member is destructed, and the `_file_impl` (shared pointer) member *of that* is also destructed. In turn, in `~posix_file_impl()`, we have a *synchronous* `::close()` syscall, with a comment saying, `Note: close() can be a blocking operation on NFS`. That appears worse than v4's `co_await ... file::close`, *but* I hope that the `_close_fut` stuff will resolve that.)

Reply:
All concerns listed here should be resolved (and documented) up to v6.